#5137: Cleanups for newer Linux distro / toolchains

olofj opened the pull request:
This is a bunch of cleanups to enable compilation on a newer distro (Debian Bookworm in my case).

In addition to this, there's a couple of small fixes to UMD needed as well, I've queued them separately in that repo.

Note that I am keeping the python dependencies outside of this PR for now, since they seem to run a bit more invasive -- especially since the dependencies will require us to uprev the import of pytorch to be based on 2.0+ instead of 1.13.

I'm also including a couple of LLVM-enabling patches from Nick, he was also instrumental in debugging earlier issues with this patch set. I'll be waiting to make sure CI runs clean before I consider pushing this up for review and merge.

```
Nicholas Smith (3):
      #5137: Fix resnet hang after aggregate initializer refactor
      #5137: enable clang-16 build + asan config
      #5137: additional clang-16 fixes after rebase

Olof Johansson (18):
      #5137: common/bfloat16: add missing include
      #5137: device.hpp: add missing include of mutex
      #5137: tt_stl/concepts.hpp: add missing pragma once
      #5137: allocator_types: add missing <functional> include
      #5137: llrt/watcher.cpp: add missing <mutex> include
      #5137: Add two missing includes of mutex and cstddef
      #5137: CoreRange: Remove designated initializers
      #5137: CoreCoord: Remove designated initializers
      #5137: ShardSpec: Remove designated initializers
      #5137: {Reader,Writer}DataMovementConfig: Use ctor instead of initializer
      #5137: third_party/fmt: Temporary fixup for newer GCC
      #5137: Use memcpy for float conversions
      #5137: Fix for possibly uninitialized variable in split_across_cores()
      #5137: tt_eager: Fix a few possibly uninitialized variable warnings
      #5137: layernorm_multi_core: Fix uninitialized variable
      #5137: setup.py: Create directory for library files
      #5137: module.mk: Don't wrap kernel setup in a python file to preserve jobserver for make
      #5137: tests: matmul_utils: fix GCC warning
```

olofj committed:
#5137: common/bfloat16: add missing include

Missing <functional> for unordered_map.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: device.hpp: add missing include of mutex

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: tt_stl/concepts.hpp: add missing pragma once

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: CoreRange: Remove designated initializers

Designated initializers for structs came over from C, and caused issues
in C++ such that it stopped being supported.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: allocator_types: add missing <functional> include

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: llrt/watcher.cpp: add missing <mutex> include

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: CoreCoord: Remove designated initializers

Same as with CoreRange, these should ideally be moved to ctors but until
then just use the ordered style initializer.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: {Reader,Writer}DataMovementConfig: Use ctor instead of initializer

Designated initializers for structs came over from C, and caused issues
in C++ such that it stopped being supported.

Since we have a ctor to use here, let's switch to that instead.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: ShardSpec: Remove designated initializers

Designated initializers for structs came over from C, and caused issues
in C++ such that it stopped being supported.

Ideally we should have a constructur here instead, but until we're there,
it's better to have ordered and non-designated initializers instead.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: third_party/fmt: Temporary fixup for newer GCC

GCC thinks the local data_[] can be overrun, maybe because it assumes
last byte should be terminating NUL? Anyway, we should uprev the fmt
library to a newer version, but until then this will take care of the
warning.

tt_metal/third_party/fmt/fmt/core.h:1998:48: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 1998 |     for (size_t i = 0; i < size; ++i) data_[i] = s[i];
      |                                       ~~~~~~~~~^~~~
tt_metal/third_party/fmt/fmt/core.h: In member function ‘constexpr decltype (ctx.begin()) fmt::v8::formatter<T, Char, typename std::enable_if<(fmt::v8::detail::type_constant<T, Char>::value != fmt::v8::detail::type::custom_type), void>::type>::parse(ParseContext&) [with ParseContext = fmt::v8::basic_format_parse_context<char>; T = fmt::v8::basic_string_view<char>; Char = char]’:
tt_metal/third_party/fmt/fmt/core.h:1990:8: note: at offset 4 into destination object ‘fmt::v8::detail::fill_t<char>::data_’ of size 4
 1990 |   Char data_[max_size] = {Char(' '), Char(0), Char(0), Char(0)};
      |        ^~~~~

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: Use memcpy for float conversions

For some reason, newer GCC refuses to treat the uint32_data variable as
initialized if reinterpret_cast<> is used. Possibly because on some
architectures the size difference of uint32_t and float means that it
could be partially uninitialized?

Anyway, the other standard pattern in C++ of using memcpy seems to not
hit the same (false) positive, so let's go with that.

Once we're on C++20, std::bit_cast will come in handy here.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: Fix for possibly uninitialized variable in split_across_cores()

ncores_per_batch_h might be uninitialized if the middle switch doesn't
hit one of the values.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: tt_eager: Fix a few possibly uninitialized variable warnings

I'll be honest, I'm not deeply familiar with the C++ rules and why the
previous constructs caused uninitialized warnings when .value() was
called in the one statement, but it did. Likely due to lifetime rules of
the statement preceding .value()?

Anyway, this makes newer GCC sufficiently happy for the code to compile.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: layernorm_multi_core: Fix uninitialized variable

Caught by newer GCC: if !b, inb_data_format remains unassigned.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: setup.py: Create directory for library files

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

arakhmati reviewed (commented):
@olofj, I just have one question. Designated initializers is a modern C++ feature. And they are very handy. They make the code easier to read and make it less error-prone. So, why are they being removed?

I have a guess as to why. They are technically only supported starting with C++20. But they do work in earlier versions. I would really like to keep them in our codebase, so can we figure out why they didn't compile for you?

olofj commented:
> @olofj, I just have one question. Designated initializers is a modern C++ feature. And they are very handy. They make the code easier to read and make it less error-prone. So, why are they being removed?
> 
> I have a guess as to why. They are technically only supported starting with C++20. But they do work in earlier versions. I would really like to keep them in our codebase, so can we figure out why they didn't compile for you?

I was actually unaware of them coming back in C++20. I just know they started causing build errors in newer GCC and C++17, since they got more pedantic.

I'd argue that they don't add clarity the way we've been using them. There's very little ambiguity in the constructs we have, and `{.x = 0, .y = 0}` vs `{0, 0}` in particular is so implicitly assumed to be the same thing that they're bad examples of the value of designated initializers.

Main reason I've seen it useful is when there's a large struct that you just want to fill a few values. But, then you end up with confusion on what fields are initialized and not, etc. This is where ctors with default values bring clarity in C++ vs C, IMHO.

So, bottom line: I don't have a preference here, I just want to be able to build tt-metal on a more modern distro than Ubuntu 20.04 myself. Getting a head of the copypasta reduplication of the current patterns will take a few iterations, and if there's interest to add it back later, that can be done.

tt-rkim commented:
The hangs on perf and device perf tests are definitely weird... I personally haven't seen such on main in a long time. and that grayskull hang looks suspicious: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7810904512

Perhaps another run, but that already seems strange to me. I'd invite thoughts from @TT-billteng and @tt-aho but I'm hesitant to call this done deal.

tt-aho commented:
Something put the device perf machine in a bad state and I had to reboot it, but I didn't have time to trace where the first hang occurred. It's suspicious that the perf and device perf were all hanging in resnet since these are different machines, so it could be an issue with your branch? I would try running it locally + rerun in regression if you confirm it doesn't seem like your changes causing the issue

olofj commented:
> Something put the device perf machine in a bad state and I had to reboot it, but I didn't have time to trace where the first hang occurred, maybe it was this branch? It's suspicious that the perf and device perf were all hanging in resnet on your pipeline since these are different machines, so it could be an issue with your branch? I would try running it locally + rerun in regression if you confirm it doesn't seem like your changes causing the issue

I can rerun locally since it's GS and I have one on the machine I've been doing this work. It could be due to data init being messed up with the initializer changes, so it's not out of the question.

I can also do better due diligence and compare compiled results before/after and audit that.

Moving this PR back into draft state to indicate no need for others to spend energy on it until I've done that. ETA is likely this weekend or so, depending on weather and spare time availability.

arakhmati commented:
> > @olofj, I just have one question. Designated initializers is a modern C++ feature. And they are very handy. They make the code easier to read and make it less error-prone. So, why are they being removed?
> > I have a guess as to why. They are technically only supported starting with C++20. But they do work in earlier versions. I would really like to keep them in our codebase, so can we figure out why they didn't compile for you?
> 
> I was actually unaware of them coming back in C++20. I just know they started causing build errors in newer GCC and C++17, since they got more pedantic.
> 
> I'd argue that they don't add clarity the way we've been using them. There's very little ambiguity in the constructs we have, and `{.x = 0, .y = 0}` vs `{0, 0}` in particular is so implicitly assumed to be the same thing that they're bad examples of the value of designated initializers.
> 
> Main reason I've seen it useful is when there's a large struct that you just want to fill a few values. But, then you end up with confusion on what fields are initialized and not, etc. This is where ctors with default values bring clarity in C++ vs C, IMHO.
> 
> So, bottom line: I don't have a preference here, I just want to be able to build tt-metal on a more modern distro than Ubuntu 20.04 myself. Getting a head of the copypasta reduplication of the current patterns will take a few iterations, and if there's interest to add it back later, that can be done.

Okay, sounds good. I want us to be able to run on the more modern distros too. So, let's merge it in and hopefully we can upgrade to C++20 soon

nsmithtt commented:
Trying to compile metal with clang-16, this PR handles ~80% of the build issues so this is awesome.

> I have a guess as to why. They are technically only supported starting with C++20. But they do work in earlier versions. I would really like to keep them in our codebase, so can we figure out why they didn't compile for you?

The issue is that designated initializers are a C99 feature and it doesn't seem like GCC enforces not being able to use them in C++.

> I'd argue that they don't add clarity the way we've been using them. There's very little ambiguity in the constructs we have, and {.x = 0, .y = 0} vs {0, 0} in particular is so implicitly assumed to be the same thing that they're bad examples of the value of designated initializers.

I second this, I've largely considered designated initializers in C++ an anti-pattern (not in C), because the second the class gets a constructor or participates in inheritance it's just too easy to end up w/ a class in an undefined state.

Take `ReaderDataMovementConfig` for example, I've starred the cases below which which violates C++ spec (https://en.cppreference.com/w/cpp/language/aggregate_initialization):
> - class type (typically, struct or union), that has
>   - ** no user-declared constructors
>   - no user-provided, inherited, or explicit constructors
>   - no user-declared or inherited constructors
>   - no private or protected direct(since C++17) non-static data members
>   - ** no base classes
>   - no virtual base classes
>   - no private or protected direct base classes
>   - no virtual member functions
>   - ** no default member initializers

nsmithtt committed:
#5137: enable clang-16 build + asan config

From Nicholas Smith <nsmith@tenstorrent.com>

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

nsmithtt committed:
#5137: additional clang-16 fixes after rebase

From Nicholas Smith <nsmith@tenstorrent.com>

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: module.mk: Don't wrap kernel setup in a python file to preserve jobserver for make

The python script is just a very thin wrapper, might as well open code
it in the makefile so that the recursive make gets the jobserver
(parallelism) working as expected.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: tests: matmul_utils: fix GCC warning

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: Add two missing includes of mutex and cstddef

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj committed:
#5137: add back set_up_kernels.py

Some of our build infra apparantly (surprise!) needs this script, so
let's keep it in until we can clean that up.

olofj commented:
Latest run: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7863459092/job/21454113604

Currently hangs/fails at:
 - [build-and-test-measure-perf / build-and-test-measure-perf (grayskull, model-runner-grayskull, virtual_machine)](https://github.com/tenstorrent-metal/tt-metal/actions/runs/7863459092/job/21454113604#logs):
   +  `models/demos/resnet/tests/test_perf_resnet.py::test_perf_virtual_machine[1-0.015-30]`
 - [build-and-test-measure-perf / build-and-test-measure-perf (grayskull, perf-grayskull, self-reset, bare_metal)](https://github.com/tenstorrent-metal/tt-metal/actions/runs/7863459092/job/21454114215#logs):
   + `models/demos/resnet/tests/test_perf_resnet.py::test_perf_bare_metal[1-0.015-25]`
 - [build-and-unit-tests-fast-dispatch / build-and-unit-tests-fast-dispatch (grayskull, grayskull)](https://github.com/tenstorrent-metal/tt-metal/actions/runs/7863459092/job/21454114465#logs):
   + `tests/tt_eager/python_api_testing/unit_testing/test_optimized_conv_multi_core.py::test_run_optimized_conv[64-32-16-16-1-1-1-1-0-0-1-1-4-2-2-1-False-False-False]`
 - [run-profiler-regression / run-profiler-regression (wormhole_b0, wormhole_b0, multi-chip-num-pcie-1, multi-chip-num-chips-2)](https://github.com/tenstorrent-metal/tt-metal/actions/runs/7863459092/job/21454115763#logs):
   + `tests/tt_metal/tools/profiler/test_device_profiler.py::test_custom_cycle_count`
   + `tests/tt_metal/tools/profiler/test_device_profiler.py::test_full_buffer`

The first 3 are GS so I can easily run them locally. The WH_B0 I need to run on a work machine, but that shouldn't be too bad either.

I'm guessing it's more issues like the ones fixed yesterday with @nsmithtt's assistance. Reproducing and doing more bisects now.

olofj committed:
#5137: Fix hangs after aggregate initializer refactor

The fix for now seems to be to consistently use the ctor instead of
opencoding the initializers.

It's not obvious why this makes a difference at this time, but we had
persistent errors without these changes. Neither LLVM, GCC nor LLVM+ASAN
seems to have complained about variable lifetimes (stack vs heap), but
still there was some significant difference between the two patterns.

Original fix by Nicholas Smith <nsmith@tenstorrent.com>, expanded by me.

Signed-off-by: Olof Johansson <olofj@tenstorrent.com>

olofj commented:
Last run passed green! There were some infra issues on wormhole that a rerun took care of: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7867635921/job/21464696675


Huge thanks to @nsmithtt for the collaboration on this. There are other things left to do, but this is the main stack of churn so it'd be great to land it.

tt-rkim commented on module.mk:
```diff
@@ -106,10 +106,10 @@ ifneq (,$(filter "$(ARCH_NAME)","wormhole" "wormhole_b0"))
 
 set_up_kernels:
-	python3 $(TT_METAL_HOME)/scripts/set_up_kernels.py --short prepare
+	$(MAKE) -f $(TT_METAL_HOME)/tt_metal/hw/Makefile-runtime -C $(TT_METAL_HOME)/tt_metal/hw prepare
```
I ideally wanted to make these scripts tested by consequence by injecting them into the build process here. But I guess this wouldn't always work

Checking functionality of these scripts should probably be something in CI/CD.

tt-rkim replied:
Yes, it's part of setting things up for the eventual install-by-wheel flow for metal that we should (and if I was dictator, I would change that to "WILL have"). Buda has similar requirements, but to my current knowledge it does that by hiding it within some `__init__.py` files at the top of tree.

There should be uses for it in the `.github/workflows/eager-package-main.yaml`.

Right now, our leadership has stressed that functionality + performance via source install are of utmost importance. So, the wheel flow has taken a backseat. The purpose of this pipeline is to guarantee we don't regress.

Please let me know your thoughts.

tt-rkim commented on tests/tt_metal/tt_metal/module.mk:
```diff
@@ -72,7 +72,7 @@ TT_METAL_TESTS += \
 
 TT_METAL_TESTS_INCLUDES = $(TEST_INCLUDES) $(TT_METAL_INCLUDES)
-TT_METAL_TESTS_LDFLAGS = -ltt_metal -ldl -lstdc++fs -pthread -lyaml-cpp -lgtest
+TT_METAL_TESTS_LDFLAGS = -ltt_metal -ldl -lstdc++fs -pthread -lyaml-cpp -lgtest -lm
```
math lib?

olofj replied:
This came in from Nick when he enabled LLVM building. My guess is that with GCC you implicitly got it included through dependencies and LLVM was pickier. We might have more minor items like this trickling in after this main set.

tt-rkim replied:
How priority of an item is it to start running on more distros / toolchains? We have an open to create a job on Github hosted runners to start compiling and uploading metal build artifacts on more build configs and platforms: https://github.com/tenstorrent-metal/tt-metal/issues/4416

Probably going to matrix over a couple of Ubuntus, and can include other toolchains. More LLVM is also in my scope. Other archs like ARM and RISC-V... lowkey dicey but still open to it. If we could get access to some internally that could work nicely, although that would technically be a self-hosted runner. I don't believe many cloud providers, if any, have generally-available RISC-V hosts.

We could specify which we want and throw those into that issue too as a comment.

olofj replied:
I had hoped to start chipping away at this, and do it best-effort style since I want it myself. Once we've done the initial cleanup (such as this patch series), we should probably setup at least build jobs for these toolchains. Happy to keep chipping away at it.

External feedback from GS owners is that they'd like to move away from the python 3.8, etc, restrictions.

DrJessop commented on tt_metal/impl/dispatch/command_queue.hpp:
```diff
@@ -453,11 +453,6 @@ class CommandQueue
-        CommandQueue(CommandQueue&&) = default;
-        CommandQueue& operator=(const CommandQueue&) = default;
-        CommandQueue& operator=(CommandQueue&&) = default;
-
```
fyi: @TT-billteng

olofj replied:
(This was also from the LLVM 16 commit)

nsmithtt replied:
This class has `std::mutex` inside it which makes the class move only and forces you to manually write a move constructor.  Just removing these default constructors makes the build happy.

DrJessop commented on tt_metal/impl/device/device.hpp:
```diff
@@ -219,7 +220,7 @@ class Device {
     std::unique_ptr<SystemMemoryManager> sysmem_manager_;
     vector<std::unique_ptr<Program, detail::ProgramDeleter>> command_queue_programs_;
-    const uint8_t num_hw_cqs_;
+    uint8_t num_hw_cqs_;
```
Why remove const qualifier?

olofj replied:
It's from Nick's LLVM commit, I'm guessing because it's not actually const, or promoted to a non-const variable somewhere -- I'm not sure. Description was "additional clang-16 fixes after rebase".

nsmithtt replied:
This will force the class to be move only and forces you to manually write the move constructor.  I can dig up the clang compiler error if you want.

nsmithtt replied:
```
./tt_metal/impl/device/device.hpp:79:13: error: explicitly defaulted move assignment operator is implicitly deleted [-Werror,-Wdefaulted-function-deleted]
    Device& operator=(Device &&other) = default;
            ^
./tt_metal/impl/device/device.hpp:223:19: note: move assignment operator of 'Device' is implicitly deleted because field 'num_hw_cqs_' is of const-qualified type 'const uint8_t' (aka 'const unsigned char')
    const uint8_t num_hw_cqs_;
                  ^
./tt_metal/impl/device/device.hpp:79:41: note: replace 'default' with 'delete'
    Device& operator=(Device &&other) = default;
```

olofj commented:
Checked with @TT-billteng on this -- since this PR touches so many files there are a ton of reviewers marked as required. It also keeps creating new conflicts, so it's hard to align all of these.

I'll go ahead and merge it now, since it's had several reviewers approve. I'll be happy to follow up with any fixes needed based on later-arriving comments and reviews. Thanks!
